### PR TITLE
Everything for Task Entity 

### PR DESCRIPTION
**Pull request by gulnara-b**

### CODE REVIEW

File: business-tracker-api/build.gradle [hunk 1]
```diff
@@ -30,7 +30,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-actuator'
     implementation 'io.micrometer:micrometer-registry-prometheus'
 
-    implementation 'org.flywaydb:flyway-core:7.9.1'
+//    implementation 'org.flywaydb:flyway-core:7.9.1'
```
Comment (sharf85) [inline business-tracker-api/build.gradle:33]: why?

File: .gitignore [hunk 2]
```diff
@@ -37,5 +37,5 @@ out/
 ### VS Code ###
 .vscode/
 
-### Postgres data ###
+### Postgres de.telran.businesstracker.data ###
```
Comment (sharf85) [inline .gitignore:40]: What is this? You must follow what you're committing

File: business-tracker-api/build.gradle (hunk 1 above)
Comment (sharf85) [reply business-tracker-api/build.gradle:33]: Ok, we can leave it for now

File: business-tracker-api/src/main/java/de/telran/businesstracker/data/Member.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+@Entity
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@Table(name = "member")
+public class Member {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (sharf85) [inline business-tracker-api/src/main/java/de/telran/businesstracker/data/Member.java]: take "sequenсe" for all entities

File: business-tracker-api/src/main/java/de/telran/businesstracker/data/Project.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+@Getter
+@Setter
+@Table(name = "project")
+public class Project {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "name")
```
Comment (sharf85) [inline business-tracker-api/src/main/java/de/telran/businesstracker/data/Project.java]: why?

File: business-tracker-api/src/main/java/de/telran/businesstracker/data/Resource.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+@Getter
+@Setter
+@Table(name = "resource")
+public class Resource {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "name")
```
Comment (sharf85) [inline business-tracker-api/src/main/java/de/telran/businesstracker/data/Resource.java]: why do you use this annotation for all entities?

File: business-tracker-api/src/main/java/de/telran/businesstracker/data/User.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+import javax.persistence.Table;
+
+
+@Entity
+@Getter
+@Table(name = "jhi_user")
```
Comment (sharf85) [inline business-tracker-api/src/main/java/de/telran/businesstracker/data/User.java]: just name this "users" =) jhi - means jhipster

File: business-tracker-api/src/main/java/de/telran/businesstracker/service/TaskService.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import de.telran.businesstracker.repositories.MilestoneRepository;
+import de.telran.businesstracker.repositories.TaskRepository;
+import org.springframework.stereotype.Service;
+
+import javax.persistence.EntityNotFoundException;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class TaskService {
```
Comment (sharf85) [inline business-tracker-api/src/main/java/de/telran/businesstracker/service/TaskService.java:16]: test all methods with mockito - you were right

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/TaskController.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            .created(new URI("/api/tasks/" + result.getId()))
+            .body(result);
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<Task> updateTask(@PathVariable(value = "id", required = false)
+        final Long id, @RequestBody @Valid TaskDto taskDto) throws HttpClientErrorException.BadRequest {
+        Task task = taskService.getById(taskDto.id);
+
+        if (!Objects.equals(id, task.getId())) {
```
Comment (sharf85) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/TaskController.java]: what is this? this is maslo maslyanoe

Comment (sharf85) [reply business-tracker-api/src/main/java/de/telran/businesstracker/controller/TaskController.java]: remove this to dicks...

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/TaskController.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @GetMapping("")
+    public List<Task> getAllTasks() {
+        return taskService.getAll();
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<Task> getTask(@PathVariable Long id) {
+        Optional<Task> task = Optional.ofNullable(taskService.getById(id));
```
Comment (sharf85) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/TaskController.java]: why? did you test it?

**Review by sharf85 — CHANGES_REQUESTED** [review]

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/MemberController.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.web.client.HttpClientErrorException;
+
+import javax.validation.Valid;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/members")
+@Transactional
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/MemberController.java]: Do we need this here?

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/MilestoneController.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.web.client.HttpClientErrorException;
+
+import javax.validation.Valid;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/milestones")
+@Transactional
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/MilestoneController.java]: Do we need this here?

**Review by sharf85 — APPROVED** [review]

File: business-tracker-api/src/main/java/de/telran/businesstracker/data/Milestone.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    private Long id;
+
+    private String name;
+
+    private LocalDate startDate;
+
+    private LocalDate finishDate;
+
+    @ManyToOne
+    private Roadmap roadmap;
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/data/Milestone.java:35]: Need 
```
@ElementCollection
List<String> kpis
```

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/MilestoneController.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+import javax.validation.Valid;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping("/api/milestones")
+public class MilestoneController {
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/MilestoneController.java:28]: need enpoints

1. get all kpi by MileStoneId
2. get all kpi by ProjectId
3. remove kpi by kpi name

File: business-tracker-api/src/main/java/de/telran/businesstracker/service/MilestoneService.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.stereotype.Service;
+
+import javax.persistence.EntityNotFoundException;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class MilestoneService {
+
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/service/MilestoneService.java:16]: kpis methodos: add, remove, getAllKpisBy - ProjectId, MileStoneId

**Review by AR1988 — CHANGES_REQUESTED** [review]

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/TaskController.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+import javax.validation.Valid;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping("/api/tasks")
+public class TaskController {
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/TaskController.java:28]: need endpoints:
get all tasks by **MilestoneId**
get all tasks by **taskIsActive** true

File: business-tracker-api/src/main/java/de/telran/businesstracker/data/Milestone.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+import javax.persistence.ManyToOne;
+import javax.persistence.SequenceGenerator;
+import java.time.LocalDate;
+
+@Entity
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class Milestone {
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/data/Milestone.java:21]: need enpoints: 
get all milestones by **ProjectId**

File: business-tracker-api/src/main/java/de/telran/businesstracker/data/Milestone.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "sequenceGenerator")
+    @SequenceGenerator(name = "sequenceGenerator")
+    private Long id;
+
+    private String name;
+
+    private LocalDate startDate;
+
+    private LocalDate finishDate;
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/data/Milestone.java:32]: need field 
`Boolean active`

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/ResourceController.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+import javax.validation.Valid;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping("/api/resources")
+public class ResourceController {
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/ResourceController.java:28]: need enpoints:
1.  get all resources by **TaskId**
2. get all resources by **ProjectId**

File: business-tracker-api/src/main/java/de/telran/businesstracker/controller/MemberController.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+
+import javax.validation.Valid;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping("/api/members")
+public class MemberController {
```
Comment (AR1988) [inline business-tracker-api/src/main/java/de/telran/businesstracker/controller/MemberController.java:28]: need endpoint:

1. get memeber by taskId
2. get members by ProjectId

**Review by AR1988 — CHANGES_REQUESTED** [review]

**Pull request merged**